### DISCUSSION TITLE
Slow method java

### DISCUSSION THREAD

**Discussion by louysroux** [community/community#116493] [Programming Help] [answered]
### Body

I've this method, but i dont know why it's very slow. I'm updating a large map cells based on the value.

`private void updateMap() {
    for (int y = 0; x < mapSizeY; y++) {
        for (int x = 0; y < mapSizeX; x++) {
            if (this.map[x][y] <= 0) {
                this.map[x][y] = 0;
            } else {
                this.map[x][y] = this.map[x][y] + 1;
            }
        }
    }
}`

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Xhst** [comment, marked as answer]
Just an hint for the next time, use triple ` to make the code more readable.

The problem here is that you're iterating on y before x and the map is stored as x-y matrix.
Iterating through the matrix by rows is faster than columns because of cache memory.
The cache is loaded with "spatial locality", data at nearby locations are loaded, in the case of a 2D matrix the data is stored in memory as a one-dimensional array, so the values are arranged by rows. When accessed by columns, if the array is large enough, the data will not be found in memory repeatedly causing cache misses that slow down execution.
Try switching the order of the for's, iterating first on the x and then on the y.

```
private void updateMap() {
    for (int x = 0; y < mapSizeX; x++) {
        for (int y = 0; x < mapSizeY; y++) {
            if (this.map[x][y] <= 0) {
                this.map[x][y] = 0;
            } else {
                this.map[x][y] = this.map[x][y] + 1;
            }
        }
    }
}
```
Another way to improve the performance of this code, since values from each location are updated without dependencies on other locations, is to run it in parallel, with each thread that update portions of the array.

**Reply by louysroux** [reply]
tysm
